File: BrianLee/Week2/2022-09-21-Search-in-Rotated-Sorted-Array.java
class Solution {
    public int search(int[] nums, int target) {
        int start = 0;
        int end = nums.length-1;
        while(start <= end) {
            int mid = (end-start)/2+start;
            if(nums[mid] == target) return mid;
            if(mid-1 >= 0 && nums[mid-1] == target) return mid-1;
            if(mid+1 < nums.length && nums[mid+1] == target) return mid+1;
            else {
                if(mid-1 >= 0 && nums[start] <= nums[mid-1]) {
                    if(nums[start] <= target && target <= nums[mid-1]) {
                        end = mid-1;
                    } else {
                        start = mid+1;
                    }
                } else {
                    if(nums[mid] < target && target <= nums[end]) {
                        start = mid+1;
                    } else {
                        end = mid-1;
                    }
                }
            }
        }
        return -1;
    }
}

File: SungJae Hong/2022-09-21-search-in-rotated-sorted-array.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = 0
        r = len(nums) - 1
        # have to put equal sign for 1 value list.
        while l <= r:
            mid = (l + r) // 2
            if target == nums[mid]:
                return mid
            # left sorted portion
            if nums[l] <= nums[mid]:
                if target > nums[mid] or target < nums[l]:
                    l = mid + 1
                else:
                    r = mid - 1
            # right sorted portion
            else:
                if target < nums[mid] or target > nums[r]:
                    r = mid - 1
                else:
                    l = mid + 1
        return -1
